Unknown error while calculating nonequillibrium spectra with HITEMP database

Dear RADIS team,

During calculation of non-equillibrium spectra there are some errors occur. I am trying to use HITEMP database for calculations. Loading and cashing goes fine. Please find traceback of these error in spoiler below:

<details>
  <summary>Note in configuration file .radis</summary>
[HITEMP2010-CO2-work]
info = HITEMP2020 CO2 lines with TIPS-2017 for partition functions (equilibrium) and RADIS for rovibrational energies (nonequilibrium)
path = D:/Blekhshtein_phd/Files/Database_files/HITEMP/CO2/*.par
format = hitran
parfuncfmt = hapi
levelsfmt = radis

</details>

<details>
  <summary>Traceback</summary>
Traceback (most recent call last):
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\indexes\base.py", line 2895, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas\_libs\index.pyx", line 70, in pandas._libs.index.IndexEngine.get_loc
  File "pandas\_libs\index.pyx", line 101, in pandas._libs.index.IndexEngine.get_loc
  File "pandas\_libs\hashtable_class_helper.pxi", line 1675, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas\_libs\hashtable_class_helper.pxi", line 1683, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: '(0.0,0.0,0.0,1.0)'
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2020.2.3\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "D:/Blekhshtein_phd/Files/Database_files/HAPI_HITRAN/fitting_kit_spec_v_arbitrary_from_db_v1.py", line 237, in <module>
    name='treanor. fit')
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\factory.py", line 1383, in non_eq_spectrum
    self._check_noneq_parameters(vib_distribution, singleTvibmode)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\base.py", line 1544, in _check_noneq_parameters
    calc_Evib_harmonic_anharmonic=calc_Evib_harmonic_anharmonic,
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\base.py", line 1656, in _calc_noneq_parameters
    df, calc_Evib_harmonic_anharmonic=calc_Evib_harmonic_anharmonic
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\base.py", line 422, in _add_Evib123Erot
    return self._add_Evib123Erot_RADIS_cls5_harmonicanharmonic(df)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\base.py", line 1409, in _add_Evib123Erot_RADIS_cls5_harmonicanharmonic
    ] = get_Evib123_RADIS_cls5_1iso_ah(df.loc[idx], iso)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\base.py", line 1356, in get_Evib123_RADIS_cls5_1iso_ah
    df = df.groupby(by=["v1u", "v2u", "l2u", "v3u"]).apply(fillEvib123u)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\groupby\groupby.py", line 859, in apply
    result = self._python_apply_general(f, self._selected_obj)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\groupby\groupby.py", line 892, in _python_apply_general
    keys, values, mutated = self.grouper.apply(f, data, self.axis)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\groupby\ops.py", line 213, in apply
    res = f(group)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\radis\lbl\base.py", line 1334, in fillEvib123u
    r["Evib1u_h"] = energies.at[viblvl, "Evib1_h"]
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\indexing.py", line 2078, in __getitem__
    return super().__getitem__(key)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\indexing.py", line 2025, in __getitem__
    return self.obj._get_value(*key, takeable=self._takeable)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\frame.py", line 3025, in _get_value
    index = self.index.get_loc(index)
  File "D:\Blekhshtein_phd\Files\Database_files\HAPI_HITRAN\venv\lib\site-packages\pandas\core\indexes\base.py", line 2897, in get_loc
    raise KeyError(key) from err
KeyError: '(0.0,0.0,0.0,1.0)'

</details>

Message with KeyError may vary depending on wavelength or may be other calculation parameters.

Could you please help me with interpretation of this error?

P.S. Dr. Pannier mentioned on slack that there could be errors in HITEMP database, in this case - how to correct them?

Thank you in advance!

Got it. In short:  it is indeed a HITEMP labelling error as I remembered (@CorentinGrimaldi did you have this one?), but it is hidden under this other error message.  

The reason you have no integer in your level names is because there is a `nan` (unknown level) somewhere. Example : 

```python
import pandas as pd
import numpy as np

' '.join(pd.Series(np.arange(100).tolist()+[np.nan]).astype(str))
Out[34]: '0.0 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 13.0 14.0 15.0 16.0 17.0 18.0 19.0 20.0 21.0 22.0 23.0 24.0 25.0 26.0 27.0 28.0 29.0 30.0 31.0 32.0 33.0 34.0 35.0 36.0 37.0 38.0 39.0 40.0 41.0 42.0 43.0 44.0 45.0 46.0 47.0 48.0 49.0 50.0 51.0 52.0 53.0 54.0 55.0 56.0 57.0 58.0 59.0 60.0 61.0 62.0 63.0 64.0 65.0 66.0 67.0 68.0 69.0 70.0 71.0 72.0 73.0 74.0 75.0 76.0 77.0 78.0 79.0 80.0 81.0 82.0 83.0 84.0 85.0 86.0 87.0 88.0 89.0 90.0 91.0 92.0 93.0 94.0 95.0 96.0 97.0 98.0 99.0 nan'

' '.join(pd.Series(np.arange(100).tolist()).astype(str))
Out[35]: '0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99'
```

--

I could reproduce your error with a minimal test case : 

```python

```


```python

from radis import SpectrumFactory

sf = SpectrumFactory(2284, 2285,         # cm-1
                  molecule='CO2',
                  isotope='1,2,3',
                  pressure=1.01325,   # bar
                  mole_fraction=0.1,
                  path_length=1,      # cm
                  wstep=0.001)
sf.load_databank(path=['./02_2250-2500_HITEMP2010.par'], format='hitran', parfuncfmt='hapi',
                 levelsfmt='radis')
sf.non_eq_spectrum(Tvib=(900, 900, 1100), Trot=700,
                    vib_distribution='treanor')
```

```
KeyError: '(0.0,0.0,0.0,1.0)'
```

But if I edit the code above not to be in 'Treanor' distributions : 

```
sf.non_eq_spectrum(Tvib=(900, 900, 1100), Trot=700)
```

There, RADIS crashes with a nice error message : 

```
AssertionError: Evib1u=NaN in line database at index 7 corresponding to Line:
 id            2.000000e+00
iso           1.000000e+00
wav           2.279003e+03
int           1.076000e-45
A             8.428000e-01
airbrd        4.860000e-02
selbrd        6.800000e-02
El            1.245169e+04
Tdpair        5.700000e-01
Pshft         0.000000e+00
branch        1.000000e+00
jl            1.390000e+02
v1u                    NaN
v2u                    NaN
l2u                    NaN
v3u                    NaN
ru                     NaN
v1l           2.000000e+00
v2l           0.000000e+00
l2l           0.000000e+00
v3l           1.000000e+00
rl            2.000000e+00
Ia            9.842040e-01
molar_mass    4.398983e+01
ju            1.400000e+02
Eu            1.473069e+04
Evib1l        2.662000e+03
Evib2l        0.000000e+00
Evib3l        2.349160e+03
Evib1u                 NaN
Evib2u                 NaN
Evib3u                 NaN
Evibu                  NaN
Evibl         5.011160e+03
Erotu                  NaN
Erotl         7.440529e+03
Name: 7, dtype: float64
```

which comes from this unlabelled HITEMP line (see `v1u=NaN` ?)


1.  For some reason I cannot explain, the nicer error doesn't show in Treanor mode. Probably a different try/except mechanism. 

2. For what matters to you : you could artificially remove this line from the line database. The new HITEMP won't be released before a few months so we could add a way to handle it directly in RADIS. I'm quite reluctant with this idea because I don't like editing physical properties withouts users knowing about it. But it could be a nice error message explaning what to do and how to remove/label the faulty lines

--- a/radis/lbl/base.py
+++ b/radis/lbl/base.py
@@ -291,9 +291,10 @@ class BaseFactory(DatabankLoader):
     # calc_einstein_coefficients
     # =========================================================================
 
-        def assert_no_nan(self, df, column):
-        """Assert there are no nan in the column, and crash with a nice
-        explanation if it is found"""
+    def assert_no_nan(self, df, column):
+        """Assert there are no nan in the column.
+
+        Crash with a nice explanation if one is found"""
         from radis.misc.printer import get_print_full
 
         try:

--- a/radis/lbl/base.py
+++ b/radis/lbl/base.py
@@ -291,21 +291,27 @@ class BaseFactory(DatabankLoader):
     # calc_einstein_coefficients
     # =========================================================================
 
-    @staticmethod
-    def assert_no_nan(df, column):
+    def assert_no_nan(self, df, column):
         """Assert there are no nan in the column, and crash with a nice
-        explanation if it is found."""
+        explanation if it is found"""
         from radis.misc.printer import get_print_full
 
         try:
             assert np.isnan(df[column]).sum() == 0
         except AssertionError as err:
             index = np.isnan(df[column]).idxmax()
+            if self.input.molecule == 'CO2':
+                fix_idea = "If using HITEMP2010 for CO2, some lines are " \
+                           "unlabelled and therefore cannot be used at " \
+                           "equilibrium. This is a known issue of the HITEMP " \
+                           "database and will soon be fixed in the " \
+                           "edition. In the meantime you can use: " \
+                           "'sf.df0.dropna(subset=['v1u'], inplace=True)' " \
+                           "where 'sf' is SpectrumFactory object"
             raise AssertionError(
                 "{0}=NaN in line database at index {1}".format(column, index)
-                + " corresponding to Line:\n {1}".format(
-                    index, get_print_full(df.loc[index])
-                )
+                + " corresponding to Line:\n {1}".format(index, \
+                get_print_full(df.loc[index]) + fix_idea)
             ) from err
 
     def _add_EvibErot(self, df, calc_Evib_harmonic_anharmonic=False):

--- a/radis/lbl/base.py
+++ b/radis/lbl/base.py
@@ -291,7 +291,7 @@ class BaseFactory(DatabankLoader):
     # calc_einstein_coefficients
     # =========================================================================
 
-    def assert_no_nan(self, df, column):
+        def assert_no_nan(self, df, column):
         """Assert there are no nan in the column, and crash with a nice
         explanation if it is found"""
         from radis.misc.printer import get_print_full
@@ -300,18 +300,18 @@ class BaseFactory(DatabankLoader):
             assert np.isnan(df[column]).sum() == 0
         except AssertionError as err:
             index = np.isnan(df[column]).idxmax()
-            if self.input.molecule == 'CO2':
-                fix_idea = "If using HITEMP2010 for CO2, some lines are " \
-                           "unlabelled and therefore cannot be used at " \
-                           "equilibrium. This is a known issue of the HITEMP " \
-                           "database and will soon be fixed in the " \
-                           "edition. In the meantime you can use: " \
-                           "'sf.df0.dropna(subset=['v1u'], inplace=True)' " \
-                           "where 'sf' is SpectrumFactory object"
+            if self.input.molecule == "CO2":
+                fix_idea = (
+                    "If using HITEMP2010 for CO2, some lines are unlabelled and therefore cannot be used at "
+                    "equilibrium. This is a known issue of the HITEMP database and will soon be fixed in the "
+                    "edition. In the meantime you can use: 'sf.df0.dropna(subset=['v1u'], inplace=True)' "
+                    "where 'sf' is SpectrumFactory object"
+                )
             raise AssertionError(
                 "{0}=NaN in line database at index {1}".format(column, index)
-                + " corresponding to Line:\n {1}".format(index, \
-                get_print_full(df.loc[index]) + fix_idea)
+                + " corresponding to Line:\n {1}".format(
+                    index, get_print_full(df.loc[index]) + fix_idea
+                )
             ) from err
 
     def _add_EvibErot(self, df, calc_Evib_harmonic_anharmonic=False):

